hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdiff/RocksDBCheckpointDiffer.java

```diff
@@ -939,6 +978,114 @@ private void populateCompactionDAG(List<String> inputFiles,
 
   }
 
+  /**
+   * This is the task definition which is run periodically by the service
+   * executor at fixed delay.
+   * It looks for snapshots in compaction DAG which are older than the allowed
+   * time to be in compaction DAG and removes them from the DAG.
+   */
+  public void pruneOlderSnapshotsWithCompactionHistory() {
+    String snapshotDir = null;
+    long currentTimeMillis = System.currentTimeMillis();
+
+    while (!snapshots.isEmpty() &&
+        (currentTimeMillis - snapshots.peek().getLeft())
+            > maxAllowedTimeInDag) {
+      snapshotDir = snapshots.poll().getRight();
+    }
+
+    if (snapshotDir != null) {
+      pruneSnapshotFileNodesFromDag(snapshotDir);
+    }
+  }
+
+  /**
+   * Prunes forward and backward DAGs when oldest snapshot with compaction
+   * history gets deleted.
+   */
+  public void pruneSnapshotFileNodesFromDag(String snapshotDir) {
+    Set<String> snapshotSstFiles = readRocksDBLiveFiles(snapshotDir);
+    if (snapshotSstFiles.isEmpty()) {
+      LOG.info("Snapshot '{}' doesn't have any sst file to remove.",
```

Comment:
This case is weird enough that might deserve a `warn` or `error`. Possibly this DB checkpoint directory is corrupted or tampered with.

```suggestion
      LOG.error("Snapshot '{}' DB checkpoint doesn't have any SST file.",
```